include/alpaka/rand/Philox/PhiloxSingle.hpp

```diff
@@ -0,0 +1,162 @@
+             *
+             * @tparam TAcc Accelerator type as defined in alpaka/acc
+             * @tparam TParams Basic parameters for the Philox algorithm
+             */
+            template<typename TAcc, typename TParams>
+            class PhiloxSingle : public traits::PhiloxBaseTraits<TAcc, TParams, PhiloxSingle<TAcc, TParams>>::Base
+            {
+            public:
+                using Trait =
+                    typename traits::PhiloxBaseTraits<TAcc, TParams, PhiloxSingle<TAcc, TParams>>; ///< Specialization
+                                                                                                   ///< for different
+                                                                                                   ///< backends
+                using Counter = typename Trait::Counter; ///< Backend-dependent Counter type
+                using Key = typename Trait::Key; ///< Backend-dependent Key type
+                using State = PhiloxStateSingle<Counter, Key>; ///< Backend-dependent State type
+
+                State state; ///< Internal engine state
+
+            protected:
+                /** Advance internal counter to the next value
+                 *
+                 * Advances the full internal counter array, resets the position pointer and stores the intermediate
+                 * result to be recalled when the user requests a number.
+                 */
+                ALPAKA_FN_HOST_ACC void advanceState()
+                {
+                    this->advanceCounter(state.counter);
+                    state.result = this->nRounds(state.counter, state.key);
+                    state.position = 0;
+                }
+
+                /** Get the next random number and advance internal state
+                 *
+                 * The intermediate result stores N = TParams::counterSize numbers. Check if we've already given out
+                 * all of them. If so, generate a new intermediate result (this also resets the pointer to the position
+                 * of the actual number). Finally, we return the actual number.
+                 *
+                 * @return The next random number
+                 */
+                ALPAKA_FN_HOST_ACC auto nextNumber()
+                {
+                    state.position++;
+                    if(state.position == TParams::counterSize)
+                    {
+                        advanceState();
+                    }
+#ifdef PLAIN_GPU_VECTOR_TYPES
+                    switch(state.position)
+                    {
+                    case 0:
+                        return state.result.x;
+                    case 1:
+                        return state.result.y;
+                    case 2:
+                        return state.result.z;
+                    case 3:
+                        return state.result.w;
+                    default: // Should never get here
+                        return 0u;
+                    }
```

Comment:
My crossing my fingers for your optimizer! That could be really inefficient. I generally wondered, why you use the vector types in CUDA an not just something like `std::array` but for CUDA.

Reply:
Well, we talked about it - std::array doesn't really work with CUDA, and this is just a tiny array with 4 numbers in it, so device arrays would be an overkill. I've looked into the cuRAND sources, and they seem to use the unit4 type as well.